mono/profiler/arg-parsing.c

```diff
@@ -0,0 +1,406 @@
+
+	for (p = desc; *p; p++){
+		switch (*p){
+		case ',':
+			if (!in_quotes) {
+				if (buffer_pos != 0){
+					buffer [buffer_pos] = 0;
+					parse_arg (buffer, config);
+					buffer_pos = 0;
+				}
+			} else {
+				buffer [buffer_pos++] = *p;
+			}
+			break;
+
+		case '\\':
+			if (p [1]) {
+				buffer [buffer_pos++] = p[1];
+				p++;
+			}
+			break;
+		case '\'':
+		case '"':
+			if (in_quotes) {
+				if (quote_char == *p)
+					in_quotes = FALSE;
+				else
+					buffer [buffer_pos++] = *p;
+			} else {
+				in_quotes = TRUE;
+				quote_char = *p;
+			}
+			break;
+		default:
+			buffer [buffer_pos++] = *p;
+			break;
+		}
+	}
+		
+	if (buffer_pos != 0) {
+		buffer [buffer_pos] = 0;
+		parse_arg (buffer, config);
+	}
+
+	//Compure config effective mask
+	config->effective_mask = config->enable_mask & ~config->disable_mask;
+}
+
+static void
+set_hsmode (ProfilerConfig *config, const char* val)
+{
+	char *end;
+	unsigned int count;
+	if (!val)
+		return;
+	if (strcmp (val, "ondemand") == 0) {
+		config->hs_mode_ondemand = TRUE;
+		return;
+	}
+
+	count = strtoul (val, &end, 10);
+	if (val == end) {
+		usage ();
+		return;
+	}
+
+	if (strcmp (end, "ms") == 0)
+		config->hs_mode_ms = count;
+	else if (strcmp (end, "gc") == 0)
+		config->hs_mode_gc = count;
+	else
+		usage ();
+}
+
+static void
+set_sample_freq (ProfilerConfig *config, const char *val)
+{
+	if (!val)
+		return;
+
+	const char *p = val;
+
+	// Is it only the frequency (new option style)?
+	if (isdigit (*p))
+		goto parse;
+
+	// Skip the sample type for backwards compatibility.
+	while (isalpha (*p))
+		p++;
+
+	// Skip the forward slash only if we got a sample type.
+	if (p != val && *p == '/') {
+		p++;
+
+		char *end;
+
+	parse:
+		config->sample_freq = strtoul (p, &end, 10);
+
+		if (p == end) {
+			usage ();
+			return;	
+		}
+
+		p = end;
+	}
+
+	if (*p)
+		usage ();
+}
+
+static void
+usage (void)
+{
+	printf ("Log profiler version %d.%d (format: %d)\n", LOG_VERSION_MAJOR, LOG_VERSION_MINOR, LOG_DATA_VERSION);
+	printf ("Usage: mono --profile=log[:OPTION1[,OPTION2...]] program.exe\n");
+	printf ("Options:\n");
+	printf ("\thelp                 show this usage info\n");
+	printf ("\t[no]'event'          enable/disable a profiling event. Valid values: domain, assembly, module, class, jit, exception, gcalloc, gc, thread, monitor, gcmove, gcroot, context, finalization, counter, gchandle\n");
+	printf ("\t[no]typesystem       enable/disable typesystem related events such as class and assembly loading\n");
+	printf ("\t[no]alloc            enable/disable recording allocation info\n");
+	printf ("\t[no]calls            enable/disable recording enter/leave method events\n");
+	printf ("\t[no]legacy           enable/disable pre mono 5.4 default profiler events\n");
+	printf ("\tsample[=TYPE]        enable/disable statistical sampling of threads (by default cycles/100)\n");
+	printf ("\t                     TYPE: cycles,instr,cacherefs,cachemiss,branches,branchmiss\n");
```

Comment:
We should only document `cycles`. The others were part of the Linux perf support which we've removed.